### ISSUE TITLE
"Unset" value for google_compute_region_instance_group_manager target_size argument?

### ISSUE THREAD

**Issue by danisla** [terraform-providers/terraform-provider-google#667]
Related issues: #65, #13. [Terraform #5471](https://github.com/hashicorp/terraform/issues/5471)

Affected resources:
 - google_compute_region_instance_group_manager
 - google_compute_instance_group_manager

The [`target_size`](https://www.terraform.io/docs/providers/google/r/compute_instance_group_manager.html#target_size) argument is mutually exclusive from the autoscaler resource. Is there a way to conditionally set the `target_size` argument so that when I'm using a autoscaler, it can be unset? 

Setting it to`0` forces my instance group to `0`, which conflicts with the autoscaler.

```
target_size = "${var.autoscaling ? 0 : var.size}"
```

If `0` was the unset value, then the above indented use case would work. If `0` is valid, can we use `-1`?

**Comment by danawillow** [comment]
Taking a look at what the cloud console claims a REST request for MIG with autoscaling is, I think the "unset" value is actually 1. Can you give that a try @danisla and see if that works?

And now it makes sense to me why 1 was originally the default here. I think the right answer for the future is to get rid of the `Computed` on it and instead set a default of 1, but leave in the `ForceSendFields` so people can set zero that want to. @rileykarson, if you have a moment, do you remember off the top of your head whether that was something you had tried? I just gave it a quick go and it looks like tests pass just fine.

**Comment by danisla** [comment]
@danawillow, I set the `target_size = 1` and autoscaling `min_replicas = 3` and `terraform plan` wants to change the `target_size` from `3` back to `1`

```
Terraform will perform the following actions:

  ~ module.mig1.google_compute_region_instance_group_manager.default
      target_size: "3" => "1"
```

I get an error if I try to apply it:

```
Error: Error applying plan:

1 error(s) occurred:

* module.mig1.google_compute_region_instance_group_manager.default: 1 error(s) occurred:

* google_compute_region_instance_group_manager.default: Error resizing RegionInstanceGroupManager: googleapi: Error 400: The instance_group_manager resource 'projects/disla-terraform-admin-101/regions/us-central1/instanceGroupManagers/autoscale-cluster' is already being used by 'projects/disla-terraform-admin-101/regions/us-central1/autoscalers/autoscale-cluster', resourceInUseByAnotherResource
```

I have a hack workaround now, which forces the `target_size` to follow the `min_replicas`, I think this will only work if the cluster is at it's minimum scale though.

```hcl
target_size = "${var.autoscaling ? var.min_replicas : var.size}"
```

**Comment by danawillow** [comment]
Ah ok, that makes sense, and now I see why this was so tricky. We want `target_size` to be computed so that if it's unset it can just go ahead and change. But we also want to be able to specify a size of zero. Unfortunately, terraform can't currently tell the difference between unset and set to zero, so I don't think there's anything I can do in the google provider at the moment to fix this.

@apparentlymart you might be interested in this since I know you're working on improvements in this area.

**Comment by danawillow** [comment]
I haven't actually checked, but this might be fixable now either with GetOkExists or CustomDiff. I'm not working on it right now though so unassigning.

**Comment by paddycarver** [comment]
This should be resolved by the [upcoming 0.12 changes](https://www.hashicorp.com/blog/terraform-0-12-conditional-operator-improvements#conditionally-omitted-arguments).

**Comment by rileykarson** [comment]
This should be possible in `0.12` with `null`.
